doc/developer/design/20250707_persist_schemas.md

```diff
@@ -0,0 +1,118 @@
+
+## Success Criteria
+
+- Migrations need to be backwards compatible: old data should remain readable when we change the schema.
+- Migrations should be metadata-only. Changing the schema of a shard should not require synchronously rewriting it; we should be able to lazily adapt the schema of old data at read or compaction time.
+- Migrations need to behave well with compaction.
+    - Compacting batches from before and after a migration should not fail.
+    - Compaction should preserve the ordering of the input data when it can.
+    - Compaction should remain “invisible” to readers: reads should get semantically-equivalent results both before and after compaction runs.
+
+## Out of Scope
+
+- We do not support arbitrary migrations. (Truly complex migrations, like rewriting columns using custom logic, will need to be supported at a higher level.)
+- We specify only the migration logic that Persist uses for its internal data representation. (ie. at the `DataType` level, not the `RelationDesc` level.)
+- Migrations do not need to be *forwards* compatible: changing the schema may cause readers with old, incompatible schemas to error out.
+    - However, it’s important that these readers detect the issue and fail cleanly, instead of returning incorrect results.
+
+## Solution Proposal
+
+> This is partly a backwards-looking design, since the initial work was done last year. This section sketches out a consistent general design… and the MVP section below covers what parts of this have been implemented already and the simplifying assumptions we’ve made.
+>
+
+### Schemas and DataTypes
+
+As part of the columnar migration, Persist introduced a new `Schema` trait which controls how a particular set of rows should be tranlated into a columnar format. (For example, the key type of an typical shard is `SourceData` and its schema type is `RelationDesc`; the arrow datatype will generally be a struct, with fields that vary depending on the exact set of columns and values in that schema.) Persist uses schema implementations like this to map inputs and outputs to and from its columnar representation at the edges, but otherwise works internally in terms of the underlying Arrow datatypes. This allows Persist to strongly enforce its internal requirements around data ordering and encoding, while allowing relations and their encoding to vary over time.
+
+A Persist shard maintains an ordered sequence of schemas. This sequence is appended to with the `PersistClient::compare_and_evolve_schema` method. If the sequence is empty, this sets the initial schema of the shard. If the sequence is non-empty, this will check that the schema obeys the schema-migration rules; if it does, it will add the given schema as the latest schema of the shard. A monotonically-increasing `SchemaId` identifies a particular schema in this sequence.
```

Comment:
> If the sequence is empty, this sets the initial schema of the shard.

Does this match the current implementation? `compare_and_evolve_schema` takes a `SchemaId` to compare the current schema to, so it seems like if there is no schema, it would be impossible to provide that `SchemaId`.

Reply:
The `register_schema` method (used internally / implicitly by writers) handles the other case - it registers the given schema if there aren't any existing schemas.

So: it does not match, but it has the ~same effect. Overall the design in this section is "cleaner" and more general than what's implemented (re. the section header) but I can see that being confusing when I'm eg. calling out specific methods... I can either add the caveat here or make this less specific.